File: legacy/src/main/rs/emulate/legacy/config/kit/Part.java
package rs.emulate.legacy.config.kit;

import java.util.Arrays;
import java.util.NoSuchElementException;

import rs.emulate.shared.util.DataBuffer;
import rs.emulate.util.Assertions;

/**
 * The body part an {@link IdentikitDefinition} is for.
 *
 * @author Major
 */
enum Part {

	/**
	 * The null Part, used as the default.
	 */
	NULL(-1),

	/**
	 * The head Part.
	 */
	HEAD(0),

	/**
	 * The chin Part.
	 */
	CHIN(1),

	/**
	 * The chest Part.
	 */
	CHEST(2),

	/**
	 * The arms Part.
	 */
	ARMS(3),

	/**
	 * The hands Part.
	 */
	HANDS(4),

	/**
	 * The legs Part.
	 */
	LEGS(5),

	/**
	 * The feet Part.
	 */
	FEET(6);

	/**
	 * The id offset for female characters.
	 */
	private static final int FEMALE_ID_OFFSET = 7;

	/**
	 * Decodes a Part from the specified {@link DataBuffer}.
	 *
	 * @param buffer The Buffer.
	 * @return The Part.
	 */
	public static Part decode(DataBuffer buffer) {
		return Part.valueOf(buffer.getUnsignedByte());
	}

	/**
	 * Encodes the specified Part into the specified {@link DataBuffer}.
	 *
	 * @param buffer The Buffer.
	 * @param part The Part.
	 */
	public static void encode(DataBuffer buffer, Part part) {
		buffer.putByte(part.id);
	}

	/**
	 * Gets the Part with the specified id.
	 *
	 * @param value The id.
	 * @return The Part.
	 * @throws NoSuchElementException If no Part with the specified id exists.
	 */
	public static Part valueOf(int value) {
		Assertions.checkWithin(0, 13, value, "Part value must be [0, 13] (received " + value + ").");
		Part[] values = values();
		int id = value % values.length; // 0-6 are male, 7-13 are female.

		return Arrays.stream(values).filter(part -> part.id == id).findAny().get();
	}

	/**
	 * The id of this Part.
	 */
	private final int id;

	/**
	 * Creates the Part.
	 *
	 * @param id The id of the Part.
	 */
	Part(int id) {
		this.id = id;
	}

	/**
	 * Gets the id of this Part for female characters.
	 *
	 * @return The id.
	 */
	public int getFemaleId() {
		return id + FEMALE_ID_OFFSET;
	}

	/**
	 * Gets the id of this Part for male characters.
	 *
	 * @return The id.
	 */
	public int getMaleId() {
		return id;
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/config/kit/Part.kt
package rs.emulate.legacy.config.kit

import rs.emulate.shared.util.getUnsignedByte
import rs.emulate.shared.util.putByte
import java.nio.ByteBuffer

/**
 * The body part an [IdentikitDefinition] is for.
 */
enum class Part(val maleId: Int) {

    /**
     * The null Part, used as the default.
     */
    NULL(-1),

    /**
     * The head Part.
     */
    HEAD(0),

    /**
     * The chin Part.
     */
    CHIN(1),

    /**
     * The chest Part.
     */
    CHEST(2),

    /**
     * The arms Part.
     */
    ARMS(3),

    /**
     * The hands Part.
     */
    HANDS(4),

    /**
     * The legs Part.
     */
    LEGS(5),

    /**
     * The feet Part.
     */
    FEET(6);

    /**
     * Gets the id of this Part for female characters.
     */
    val femaleId: Int = maleId + FEMALE_ID_OFFSET

    companion object {

        /**
         * Decodes a Part from the specified [ByteBuffer].
         */
        fun decode(buffer: ByteBuffer): Part {
            return Part.valueOf(buffer.getUnsignedByte())
        }

        /**
         * Encodes the specified Part into the specified [ByteBuffer].
         */
        fun encode(buffer: ByteBuffer, part: Part): ByteBuffer {
            return buffer.putByte(part.maleId)
        }

        /**
         * Gets the Part with the specified id.
         */
        fun valueOf(value: Int): Part {
            require(value in 0..13) { "Part value must be [0, 13] (received $value)." }
            val values = values()
            val id = value % (values.size - 1) // 0-6 are male, 7-13 are female.

            return values.find { part -> part.maleId == id }!!
        }
    }

}

/**
 * The id offset for female characters.
 */
private const val FEMALE_ID_OFFSET = 7
